cpp/src/gandiva/gdv_string_function_stubs.cc

```diff
@@ -264,6 +264,356 @@ const char* gdv_fn_lower_utf8(int64_t context, const char* data, int32_t data_le
   return out;
 }
 
+//converts utf8 to utf16 encoding
+GANDIVA_EXPORT
+const char* gdv_fn_encode(int64_t context, const char* data, int32_t data_len,
+                          const char* charset, int32_t charset_len, int32_t* out_len) { 
+  if (data_len == 0) {
+    *out_len = 0;
+    return "";
+  }
+  
+  char* out = reinterpret_cast<char*>(
+    gdv_fn_context_arena_malloc(context, 32 * data_len));
+  if (out == nullptr) {
+    gdv_fn_context_set_error_msg(context,
+    "Could not allocate memory for output string");
+    *out_len = 0;
+    return "";
+  }
+
+  int32_t char_len;
+  uint32_t char_codepoint;
+  
+
+  if (memcmp(charset, "UTF-16BE", 8) == 0) {
+    char *tmp=out;
+
+    for (int32_t i = 0; i < data_len; i += char_len) {
+      
+      
+      char_len = gdv_fn_utf8_char_length(data[i]);
+
+      if (char_len == 1) {
+            
+            int32_t input = static_cast<int32_t>(data[i]);
+
+            char hex_string[20];
+            sprintf(hex_string, "%.4X", input);
+
+            memcpy(tmp, hex_string, 4);
+            tmp+=4;
```

Comment:
I think you can use the snprintf function directly in the output buffer instead of creating a temporary one and calling memcpy. I will put the logic below:
```cpp
auto index_written = 0

auto num_bytes = sprintf(hex_string + index_written, "%.4X", input);

index_written += num_bytes;
```

Reply:
We used the logic to make the output buffer of the hash function: https://github.com/apache/arrow/blob/3d6240c1ee7802829d2ed209f4135906e9413915/cpp/src/gandiva/hash_utils.cc#L129